### DISCUSSION TITLE
Does Ocelot support sharing routes and rate limit configurations between its instances if we deploy more than one Ocelot instance?

### DISCUSSION THREAD

**Discussion by MMoussaSURE** [ThreeMammals/Ocelot#2098] [Q&A] [answered]
## Expected Behavior / New Feature
I want to publish two instance of Ocelot at different servers but the under the same domain 
and if the user called my domain `/{upstreamRoute}` one of Ocelot instances should achieve the request but also the two instances should  achieve the rate limit of any one of them called, for example:
```json
  "Routes": [
          {
            "DownstreamPathTemplate": "/product",
            "DownstreamScheme": "https",
             "DownstreamHostAndPorts": [
               {
                 "Host": "products.com",
                 "Port": 44322
               }
             ],
             "UpstreamPathTemplate": "/api/product",
             "UpstreamHttpMethod": [ "POST" ],
             "RouteIsCaseSensitive": true,
             "RateLimitOptions": {
                      "ClientWhitelist": [],
                       "EnableRateLimiting": true,
                       "Period": "60s",
                       "PeriodTimespan": 60,
                       "Limit": 1
                     } 
               }
]
"GlobalConfiguration": {
  "BaseUrl": "https://api.products",
  "RateLimitOptions": {
    "DisableRateLimitHeaders": false,
    "QuotaExceededMessage": "Too Many Requests At Short Duration!",
    "HttpStatusCode": 429
  }
}
```
What should happen is 
  - if there is more than one  request reach my Ocelot domain during current 60s
  - one of ocelot instance should achieve the request and also at the same time the two ocelot instances should know that the remaining  allowed request is 0

**Comment by raman-m** [comment, marked as answer]
The short answer is **NO**.

Each Ocelot instance counts requests independently for rate limiting due to its internal cache storage. However, if you configure your instance to use a distributed cache, then the rate limiting counters will store values in this distributed cache storage.

Therefore, you should use a distributed cache, as Ocelot does not have the capability to allow one instance to share its internal cache with other instances. Nonetheless, this could be an interesting feature to explore further.

**Reply by MMoussaSURE** [reply]
thank you for your answer
but  can you show me configure my instance to use a distributed cache (redis )


for more info 
i use ocelot v : 23.2.2

**Reply by raman-m** [reply]
For how many years have you been developing software using the .NET framework?
For how many years have you been using Ocelot as a gateway?
What's your full name?
And what's your LinkedIn?

**Reply by raman-m** [reply]
Consider reading this: "[Your Own Caching](https://ocelot.readthedocs.io/en/latest/features/caching.html#your-own-caching)" section in the Ocelot documentation. However, the paragraph provided seems brief and incomplete.

**Reply by raman-m** [reply]
### 🆗 Let's research the Ocelot code together...
We have this Ocelot feature which is a part of Ocelot kernel: [AddOcelotCache](https://github.com/search?q=repo%3AThreeMammals%2FOcelot+%22AddOcelotCache%22&type=code)
So, we have these services in DI-container:
https://github.com/ThreeMammals/Ocelot/blob/8c0180a6461984ebd11ceca5544ca9d0e1564975/src/Ocelot/DependencyInjection/Features.cs#L32-L36
The 3rd and 4th services should not be changed at all!
Our goal to replace 1st and the seconds services of the `IOcelotCache<T>` type.
There is default service implementation `DefaultMemoryCache<T>` which is added to DI. This is standard In-Memory cache:
https://github.com/ThreeMammals/Ocelot/blob/8c0180a6461984ebd11ceca5544ca9d0e1564975/src/Ocelot/Cache/DefaultMemoryCache.cs#L5
This class implements target `IOcelotCache<T>` interface which is added to DI 👇 Also pay attention to `IMemoryCache` which is Microsoft interface:
https://github.com/ThreeMammals/Ocelot/blob/8c0180a6461984ebd11ceca5544ca9d0e1564975/src/Ocelot/Cache/DefaultMemoryCache.cs#L5-L14
If you possess sufficient experience with DI and MS .NET, you might consider two designs:
- Replacing the .NET framework `IMemoryCache` object
- Replacing the Ocelot `IOcelotCache<T>` object

Additionally, it's important to note that replacing either of these interfaces in DI will result in Ocelot utilizing your custom caching storage❗

**Reply by raman-m** [reply]
### The second design pertains to the storage of [Rate Limiting](https://ocelot.readthedocs.io/en/latest/features/ratelimiting.html)
The [Rate Limiting](https://ocelot.readthedocs.io/en/latest/features/ratelimiting.html) feature is incorporated into DI via the  [AddRateLimiting](https://github.com/search?q=repo%3AThreeMammals%2FOcelot%20AddRateLimiting&type=code) method.
The objective is to substitute the following interface object:
https://github.com/ThreeMammals/Ocelot/blob/8c0180a6461984ebd11ceca5544ca9d0e1564975/src/Ocelot/DependencyInjection/Features.cs#L22

Rate limiting storage relies heavily on the [IRateLimitStorage](https://github.com/search?q=repo%3AThreeMammals%2FOcelot%20IRateLimitStorage&type=code) interface for saving and reading rate limiting counters to and from persistent storage, which in practice is a memory cache: [MemoryCacheRateLimitStorage](https://github.com/ThreeMammals/Ocelot/blob/8c0180a6461984ebd11ceca5544ca9d0e1564975/src/Ocelot/RateLimiting/MemoryCacheRateLimitStorage.cs#L11) which inherits from the Microsoft `IMemoryCache` interface.

There is another undocumented class, `DistributedCacheRateLimitStorage`:
https://github.com/ThreeMammals/Ocelot/blob/8c0180a6461984ebd11ceca5544ca9d0e1564975/src/Ocelot/RateLimiting/DistributedCacheRateLimitStorage.cs#L12

This class incorporates the `Microsoft.Extensions.Caching.Distributed.IDistributedCache` object as seen in the provided code segment from the Ocelot library:
https://github.com/ThreeMammals/Ocelot/blob/8c0180a6461984ebd11ceca5544ca9d0e1564975/src/Ocelot/RateLimiting/DistributedCacheRateLimitStorage.cs#L12-L16

**Our objectives include**:
- Integrating a native `IDistributedCache` object through native storage driver helpers, see [MS docs](https://learn.microsoft.com/en-us/aspnet/core/performance/caching/distributed)
- Substituting the `IRateLimitStorage` object with an actual `DistributedCacheRateLimitStorage` object

The pseudocode could be structured as follows:
```csharp
    var builder = WebApplication.CreateBuilder(args);
    builder.Configuration.AddJsonFile("ocelot.json");
    builder.Services.AddOcelot()
        .Services.RemoveAll<Ocelot.RateLimiting.IRateLimitStorage>()
        .AddSingleton(typeof(Ocelot.RateLimiting.IRateLimitStorage), typeof(Ocelot.RateLimiting.DistributedCacheRateLimitStorage));
        // Add Microsoft.Extensions.Caching.Distributed.IDistributedCache services
```
Please be aware that incorporating the `Microsoft.Extensions.Caching.Distributed.IDistributedCache` object falls outside the scope of Ocelot's responsibilities.
As a developer, you should consult the excellent official Microsoft documentation:
- [Distributed caching in ASP.NET Core | Microsoft Learn](https://learn.microsoft.com/en-us/aspnet/core/performance/caching/distributed)

This superb guide will assist you in configuring any distributed persistent storage solutions, including _**Redis**_, SQL Server, Azure CosmosDB, and others.

### Hope it helps!

**Reply by MMoussaSURE** [reply]
> For how many years have you been developing software using the .NET framework? For how many years have you been using Ocelot as a gateway? What's your full name? And what's your LinkedIn?

i'm mohamed ragab 
net developer with 6 years of experience 
and that is first time to use ocelot 
(i write you from my second github account)
the main one is https://github.com/MohamedRagab10793-design

**Reply by MMoussaSURE** [reply]
@raman-m wrote [on Jun 14, 2024](https://github.com/ThreeMammals/Ocelot/discussions/2098#discussioncomment-9772986)
>### 🆗 Let's research the Ocelot code together...
Thank you for your clearness 
This may be useful if i want to store all configurations at distributed cache like (Redis) instead of standard cache (memory cache)

**Reply by MMoussaSURE** [reply]
@raman-m wrote [on Jun 14, 2024](https://github.com/ThreeMammals/Ocelot/discussions/2098#discussioncomment-9773605)
> ### The second design pertains to the storage of [Rate Limiting](https://ocelot.readthedocs.io/en/latest/features/ratelimiting.html)

This what I searching for.
Store rate limit values at distributed cache (Redis) to make all ocelot hosted instances have the same rate limit values at the same time.

I apply that after reading your solution throw the following 

I just add `Microsoft.Extensions.Caching.StackExchangeRedis` package version 8.0.6 and update Ocelot to version 23.3.3  
Then configure my services like
```csharp
context.Services.AddStackExchangeRedisCache(options =>
{
    options.Configuration = configuration["Redis:ConnectionString"]; 
    options.InstanceName = "BaseCacheKey:OcelotRateLimit:";
});

context.Services.AddOcelot(context.Services.GetConfiguration())
                .AddAdministration("/ocelotAdministration", "secretKey")
                .Services.RemoveAll<IRateLimitStorage>()
                         .AddSingleton(typeof(IRateLimitStorage), typeof(DistributedCacheRateLimitStorage));
```           
@raman-m Is there anything in the above configuration that is inconsistent with my needs?

**Reply by raman-m** [reply]
Good luck! Once all development is complete, it would be beneficial to demonstrate a real, working solution. Perhaps we could prepare a [sample project](https://github.com/ThreeMammals/Ocelot/tree/develop/samples) to publish in our repository.
Do you have anything else to add, Mohamed?
